### PR TITLE
add departmentId

### PR DESCRIPTION
**Pull request by VladSpitsin**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/952

### CODE REVIEW

**Review by asyagraf — APPROVED** [review]

File: src/DepartmentService.Business/Project/EditProjectCommand.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      }
+
+      OperationResultResponse<Guid?> response = new();
+
+      await _departmentProjectRepository.RemoveAsync(request.ProjectId);
+
+      if (request.DepartmentId is not null)
+      {
+        response.Body = await _departmentProjectRepository.CreateAsync(
+          _mapper.Map(
```
Comment (MariyaSin) [inline src/DepartmentService.Business/Project/EditProjectCommand.cs:56]: send the entire request to the map
you can get user id in mapper

File: src/DepartmentService.Data/DepartmentProjectRepository.cs [hunk 2]
```diff
@@ -1,47 +1,67 @@
...
+        return null;
+      }
+
+      _provider.DepartmentsProjects.Add(dbDepartmentProject);
+      await _provider.SaveAsync();
+
+      return dbDepartmentProject.Id;
+    }
+
+    public async Task<DbDepartmentUser> GetAsync(Guid userId, bool includeDepartment = false)
```
Comment (MariyaSin) [inline src/DepartmentService.Data/DepartmentProjectRepository.cs:40]: by userId???

File: src/DepartmentService.Models.Dto/Requests/EditDepartmentRequest.cs [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
+
+namespace LT.DigitalOffice.DepartmentService.Models.Dto.Requests
+{
+  public record EditDepartmentRequest
+  {
+    public string Name { get; set; }
+    public string Description { get; set; }
+    public Guid DirectorId { get; set; }
+    public bool IsActive { get; set; }
+    public Guid? DepartmentId { get; set; }
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Dto/Requests/EditDepartmentRequest.cs:11]: Why do you add it here?

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 4]
```diff
@@ -1,95 +1,113 @@
...
+      #region Paths
+
+      AddСorrectPaths(
+        new List<string>
+        {
+          nameof(EditDepartmentRequest.Name),
+          nameof(EditDepartmentRequest.Description),
+          nameof(EditDepartmentRequest.DirectorId),
+          nameof(EditDepartmentRequest.IsActive),
+          nameof(EditDepartmentRequest.DepartmentId),
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:29]: it is wrong

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 5]
```diff
@@ -1,95 +1,113 @@
...
+        });
+
+      AddСorrectOperations(nameof(EditDepartmentRequest.Name), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.Description), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.DirectorId), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.IsActive), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.DepartmentId), new List<OperationType> { OperationType.Replace });
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:36]: same

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 6]
```diff
@@ -1,95 +1,113 @@
...
+      AddFailureForPropertyIf(
+        nameof(EditDepartmentRequest.IsActive),
+        x => x == OperationType.Replace,
+        new()
+        {
+          { x => bool.TryParse(x.value?.ToString(), out bool _), "Incorrect format of IsActive." },
+        });
+
+      #endregion
+
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:90]: same

File: src/DepartmentService/Controllers/ProjectController.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+using LT.DigitalOffice.Kernel.Responses;
+using Microsoft.AspNetCore.Mvc;
+
+namespace LT.DigitalOffice.DepartmentService.Controllers
+{
+  [Route("[controller]")]
+  [ApiController]
+  public class ProjectController : ControllerBase
+  {
+    [HttpPut("edit")]
```
Comment (ndoriomedov) [inline src/DepartmentService/Controllers/ProjectController.cs:14]: Why not HttpPatch?

Comment (VladSpitsin) [reply src/DepartmentService/Controllers/ProjectController.cs:14]: Mariya say to use HttpPut

File: src/DepartmentService.Business/Project/EditProjectCommand.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      }
+
+      OperationResultResponse<Guid?> response = new();
+
+      await _departmentProjectRepository.RemoveAsync(request.ProjectId);
+
+      if (request.DepartmentId is not null)
+      {
+        response.Body = await _departmentProjectRepository.CreateAsync(_mapper.Map(request));
+        response.Status = OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/DepartmentService.Business/Project/EditProjectCommand.cs:56]: status sets by default

File: src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs [hunk 9]
```diff
@@ -1,22 +1,45 @@
...
+        Id = Guid.NewGuid(),
+        ProjectId = projectId,
+        DepartmentId = departmentId,
+        IsActive = true,
+        CreatedBy = createdBy,
+        CreatedAtUtc = DateTime.UtcNow
+      };
+    }
+
+    public DbDepartmentProject Map(EditDepartmentProjectRequest request)
```
Comment (MariyaSin) [inline src/DepartmentService.Mappers/Db/DbDepartmentProjectMapper.cs:32]: where do you use it?

**Review by actimellka — APPROVED** [review]

File: src/DepartmentService.Business/Project/EditProjectCommand.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      _mapper = mapper;
+      _departmentProjectRepository = departmentProjectRepository;
+      _departmentUserRepository = departmentUserRepository;
+      _responseCreator = responseCreator;
+    }
+
+    public async Task<OperationResultResponse<Guid?>> ExecuteAsync(EditDepartmentProjectRequest request)
+    {
+      if (!await _accessValidator.HasRightsAsync(Rights.AddEditRemoveDepartments) &&
+        !(await _accessValidator.HasRightsAsync(Rights.EditDepartmentUsers) &&
```
Comment (MariyaSin) [inline src/DepartmentService.Business/Project/EditProjectCommand.cs:45]: update kernel version and change  EditDepartmentUsers to AddRemoveDepartmentData

File: src/DepartmentService.Validation/EditDepartmentRequestValidator.cs [hunk 11]
```diff
@@ -1,95 +1,96 @@
...
+          nameof(EditDepartmentRequest.DirectorId),
+          nameof(EditDepartmentRequest.IsActive)
+        });
+
+      AddСorrectOperations(nameof(EditDepartmentRequest.Name), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.Description), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.DirectorId), new() { OperationType.Replace });
+      AddСorrectOperations(nameof(EditDepartmentRequest.IsActive), new() { OperationType.Replace });
+
+
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/EditDepartmentRequestValidator.cs:36]: remove empty line

**Review by MariyaSin — APPROVED** [review]

**Pull request merged**